### ISSUE TITLE
Specify the serializer for a certain object

### ISSUE THREAD

**Issue by suquark** [ray-project/ray#11721]
### Describe your feature request

The default serializer has many issues with certain objects. One case is the pydantic model. We can specify a flag `__ray_serializer__` in these object types to enforce ray using pickle as the default serializer instead. This avoids changing other functions like `ray.put`, `ray.remote` and `xxxx.remote(...)`. This could also cover our major cases and is super flexible.

One example could be:

```
from pydantic import BaseModel, validator

class UserModel(BaseModel):
    __ray_serializer__ = 'pickle'
    name: str
    username: str
    password1: str
    password2: str

    @validator('name')
    def name_must_contain_space(cls, v):
        if ' ' not in v:
            raise ValueError('must contain a space')
        return v.title()

    @validator('password2')
    def passwords_match(cls, v, values, **kwargs):
        if 'password1' in values and v != values['password1']:
            raise ValueError('passwords do not match')
        return v

    @validator('username')
    def username_alphanumeric(cls, v):
        assert v.isalnum(), 'must be alphanumeric'
        return v
```

**Comment by simon-mo** [comment]
@pcmoritz please take a look.

**Comment by simon-mo** [comment]
@robertnishihara raised a good point: 
> @simon-mo I have the same question as @thomasdesr I think. What happens if UserModel has __ray_serializer__ = 'pickle' and then I pass [UserModel()] into a task? presumably cloudpickle is used for the outer list, but what about the inner UserModel object? (edited) 

@suquark thoughts?

**Comment by suquark** [comment]
hmmmm, that's a good question. how about another style:

```python
with ray.serialization.serializer.pickle():
     # ... contents inside will use pickle as the serializer
```

**Comment by pcmoritz** [comment]
The with context won't work I think because the user doesn't have control over the Ray codepath that invokes the serialization. We should try to overload `__reduce__` for these objects, that should allow you to have full flexibility on how to serialize those objects. If the actual class is not serializable/too large, you can also return a function as the first return value of `__reduce__` to reinstantiate the object.

**Comment by simon-mo** [comment]
I think @pcmoritz has a great idea here:
> Hey Simon, I think you can circumvent the requirement to pickle the class for __reduce__ if you do something like this:
```
In [10]: def unpickle(*args):
    ...:     return # ...
    ...: class Foo(object):
    ...:   def __init__(self, file_name = 'file.txt'):
    ...:     self.file_name = file_name
    ...:     self.f = open(self.file_name, 'w')
    ...: 
    ...:   def __reduce__(self):
    ...:     return (unpickle, (self.file_name, ))
```
> I.e. instead of putting the class as the first argument for the return value of reduce, put a custom function.


Now I just need to define unparse as:
```
def unparse(*args):
    from package import Foo
    return Foo(*args)
```

**Comment by suquark** [comment]
yeah, a trivial solution is to overload `__reduce__` with pickle.dumps, so cloudpickle then pickles the pickled content again.
